addons/website/controllers/form.py

```diff
@@ -334,3 +337,127 @@ def insert_attachment(self, model_sudo, id_record, files):
+                message_type="comment",
+            )
+
+        return request.make_response(
+            json.dumps({"id": users.id}),
+            headers=[("Content-Type", "application/json")]
+        )
+
+    def _prepare_signup_values(self, qcontext):
+        if qcontext.get('password') != qcontext.get('confirm_password'):
+            error_data = {"error": _("Passwords do not match."), "field": "confirm_password"}
+            raise UserError(json.dumps(error_data))
+
+        values = super()._prepare_signup_values(qcontext)
+        params = dict(request.params)
+
+        # This method is also called when the user is redirected to the reset
+        # password page. In that case, we don't want to update the values.
+        if request.httprequest.path == "/web/reset_password":
+            return values
+
+        existing_fields = set(self.env["res.users"]._fields.keys())
+        filtered_params = {}
+        for key, value in params.items():
+            if isinstance(value, werkzeug.datastructures.FileStorage):
+                # Normalize key to remove any indexing (e.g., image_1920[0][0]
+                # → image_1920 / image_1024[0][0] → image_1024)
+                normalized_key = key.split('[')[0]
+                if normalized_key in existing_fields:
+                    if not value.mimetype.startswith("image/"):
+                        raise UserError(_("Only image files are allowed."))
+                    filtered_params[normalized_key] = base64.b64encode(value.read()).decode("utf-8")
+
+            elif key in existing_fields:
+                filtered_params[key] = value
+
+        values.update(filtered_params)
+        return values
+
+    def _prepare_message_for_non_existing_field(self, partner, kwargs):
+        """
+        Prepare message body and attachments for fields in kwargs that do not
+        exist on res.users.
+        Returns (message_body_parts, attachments)
+        """
+        existing_fields = set(self.env["res.users"]._fields.keys())
+        non_existing_fields = {}
+        for key, value in kwargs.items():
+            if key == "confirm_password":
+                continue
+            if isinstance(value, werkzeug.datastructures.FileStorage):
+                # Normalize key to remove index pattern (e.g.,
+                # 'image_1920[1][0]' → 'image_1920')
+                normalized_key = key.split('[')[0]
+                if normalized_key in existing_fields:
+                    continue
+            # Keep all other unknown fields
+            if key not in existing_fields:
+                non_existing_fields[key] = value
+
+        # Separate text fields and file attachments
+        attachments, text_fields = [], {}
+        for key, value in non_existing_fields.items():
+            if isinstance(value, werkzeug.datastructures.FileStorage):
+                attachment = self.env["ir.attachment"].sudo().create({
```

Comment:
We can move this code outside for loop, prepare vals and create multiple values outside loop